go/libraries/doltcore/sqle/transactions.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqle
+
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/env"
+	"github.com/dolthub/dolt/go/libraries/doltcore/merge"
+	"github.com/dolthub/dolt/go/libraries/doltcore/ref"
+	"github.com/dolthub/dolt/go/store/datas"
+	"github.com/dolthub/dolt/go/store/hash"
+)
+
+const (
+	maxTxCommitRetries = 5
+)
+
+type DoltTransaction struct {
+	startRoot  *doltdb.RootValue
+	workingSet ref.WorkingSetRef
+	db         *doltdb.DoltDB
+	rsw        env.RepoStateWriter
+}
+
+func NewDoltTransaction(startRoot *doltdb.RootValue, workingSet ref.WorkingSetRef, db *doltdb.DoltDB, rsw env.RepoStateWriter) *DoltTransaction {
+	return &DoltTransaction{
+		startRoot:  startRoot,
+		workingSet: workingSet,
+		db:         db,
+		rsw:        rsw,
+	}
+}
+
+func (tx DoltTransaction) String() string {
+	// TODO: return more info (hashes need caching)
+	return "DoltTransaction"
+}
+
+// Commit attempts to merge newRoot into the working set
+// Uses the same algorithm as merge.Merger:
+// |ws.root| is the root
+// |newRoot| is the mergeRoot
+// |tx.startRoot| is ancRoot
+// if working set == ancRoot, attempt a fast-forward merge
+func (tx *DoltTransaction) Commit(ctx *sql.Context, newRoot *doltdb.RootValue) (*doltdb.RootValue, error) {
+	for i := 0; i < maxTxCommitRetries; i++ {
+		ws, err := tx.db.ResolveWorkingSet(ctx, tx.workingSet)
+		if err == doltdb.ErrWorkingSetNotFound {
+			// initial commit
+			err = tx.db.UpdateWorkingSet(ctx, tx.workingSet, newRoot, hash.Hash{})
+			if err == datas.ErrOptimisticLockFailed {
+				continue
+			}
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		root := ws.RootValue()
+
+		hash, err := ws.Struct().Hash(tx.db.Format())
+		if err != nil {
+			return nil, err
+		}
+
+		if rootsEqual(root, tx.startRoot) {
+			// ff merge
+			err = tx.db.UpdateWorkingSet(ctx, tx.workingSet, newRoot, hash)
+			if err == datas.ErrOptimisticLockFailed {
+				continue
+			}
+		}
+
+		mergedRoot, _, err := merge.MergeRoots(ctx, root, newRoot, tx.startRoot)
```

Comment:
how are we handling merge conflicts here?

Reply:
We were not, fixed